Adds Hue sensors

robmarkcole opened the pull request:
## Description:
Replaces pull [9449](https://github.com/home-assistant/home-assistant/pull/9449
) .

Component for reading Hue motion sensors and 4 button remotes.

**Fixes:** #10383

**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io/pull/3380) with documentation (if applicable):** home-assistant/home-assistant.github.io#3380

## Example entry for `configuration.yaml` (if applicable):
```yaml
hue:

sensor:
  - platform: hue
```

## Checklist:

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**

robmarkcole committed:
Adds Hue

robmarkcole committed:
Updates CODEOWNERS

robmarkcole commented:
To add, test:

You should mock `parse_hue_api_response` with the result that it would produce from the JSON.
The actual raw JSON being converted to what it produces is part of the hue sensors package.

Also I am unclear from the docs exactly [how to](https://home-assistant.io/developers/development_checklist/) add to `requirements_all.txt`. Should I edit `script/gen_requirements_all.py` then run it?
Cheers

sdague commented:
It seems a little odd to me to do this with a brand new library instead of using phue for this. Having to have duplicate polling connections to the hue hubs seems like something we'd want to avoid, especially as hitting the API too hard can cause events to be missed.

robmarkcole commented:
HI @sdague I agree using phue would be neater - this component evolved out of a REST sensor. Indeed I was running about 6 REST sensors in place of this component, each making a request every second, and I didn't notice any performance issues. However I will look into phue, although this component is complete now (just tests to complete) and only adds a single request per second, which I don't think will be a problem given my earlier experience. Cheers

robmarkcole committed:
Update hue-sensors to v1.1

Updates to hue-sensors to v1.1 which fixes a bug with Tap sensors. Adds
requirements

robmarkcole committed:
Updates test

robmarkcole committed:
Updates test

robmarkcole commented:
Since the component performs a request I am mocking the response to the request in the test.

azogue commented:
**I'm testing this new component and I think it's great**. Loving the possibility to attach complex automations to **long presses** in the Hue dimmer switch.

Only one question: what about selecting the sensors to include? I see that the platform loads all available sensors by default (in my case, the dimmer switch and a 'phone' entity). It could use a list of `devices` to limit them, and, if not defined, load them all. What do you think?

azogue commented:
> Only one question: what about selecting the sensors to include? I see that the platform loads all available sensors by default (in my case, the dimmer switch and a 'phone' entity). It could use a list of devices to limit them, and, if not defined, load them all. What do you think?

I'm testing this new platform successfully with these changes, as a custom component, to be able to filter devices by model (and hide the undesired geofence):

```python
from homeassistant.components.sensor import PLATFORM_SCHEMA
from homeassistant.const import CONF_DEVICES

# ...

# Device icon / attrs
MODELS = ['SML', 'RWL', 'ZGP', 'GEO']
MODELS_ATTRS = {
    'SML': {'icon': 'mdi:run-fast',
            'attrs': ['light_level', 'battery', 'last_updated',
                      'lux', 'dark', 'daylight', 'temperature']},
    'RWL': {'icon': 'mdi:remote', 'attrs': ['battery', 'last_updated']},
    'ZGP': {'icon': 'mdi:remote', 'attrs': ['battery', 'last_updated']},
    'GEO': {'icon': 'mdi:cellphone', 'attrs': []}}

PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
    vol.Optional(CONF_DEVICES): cv.ensure_list(MODELS),
})

def setup_platform(hass, config, add_devices, discovery_info=None):
    # ...
    models = config.get(CONF_DEVICES)
    for key in data.data.keys():
        if not models or data.data[key]['model'] in models:
            sensors.append(HueSensor(key, data))

# ...

    def __init__(self, hue_id, data):
        # ...
        self._icon = MODELS_ATTRS[self._model]['icon']

    def update(self):
        """Update the sensor."""
        self._data.update()
        self._state = self._data.data[self._hue_id]['state']
        for attr in MODELS_ATTRS[self._model]['attrs']:
            self._attributes[attr] = self._data.data[self._hue_id][attr]
```

The yaml config to show only the dimmer switch is:
```yaml
hue:

sensor:
- platform: hue
  devices:
   - RWL
#   - GEO
```

Another change I did was changing the scan interval timedelta to 5 seconds, don't know if requesting the bridge every second was a good idea.

marthoc commented:
Related to @sdague’s comment above, would a  “cleaner” approach to this be to reorganize hue into a platform that handles the bridge, and then separate light and sensor components? So the configuration.yaml syntax would become:

```
hue:
  host:
  etc (other generic hue options):

light:
  - platform: hue
    etc (other light specific options):

sensor:
  - platform: hue
    etc (other sensor specific options, e.g. as suggested by @azogue):
```

This approach would probably line up better with how other platforms handle separate components, and, would also address better the situation for users who are not using `discovery:` since with the current approach those users would need to specify `hue:` and `host:` under the `light:` platform even if they aren’t using hue lights (I’m willing to concede this is probably an edge case, but @robmarkcole this functionality is so good it might compel users to use a hue bridge and remotes to control lots of things, who knows).

robmarkcole commented:
I agree. I think this is the preferred approach

pvizeli commented:
Please reopen this PR, if you are ready. I think like @sdague say, we should use `phue` for any kind of hue communication. If there is a good reason we can also replace `phue` but we should not use 2 python modules for the same.

robmarkcole commented:
@pvizeli happy to look at the phue approach, although first the auth needs to be split from the hue lights component. Who do you recommend approaching about that step?

andreacampi commented:
I'd be happy to refactor the existing light and split out the basic platform; I was planning on doing it in #10383 anyway :)
